akka-stream/src/main/scala/akka/stream/impl/fusing/FlatMapPrefix.scala

```diff
@@ -42,62 +45,62 @@ import scala.util.control.NonFatal
-          if (accumulated.size == n) {
-            materializeFlow()
-          } else {
-            //gi'me some more!
-            pull(in)
-          }
+        subSource match {
+          case OptionVal.Some(s) => s.push(grab(in))
+          case OptionVal.None =>
+            accumulated.append(grab(in))
+            if (accumulated.size == n) {
+              materializeFlow()
+            } else {
+              //gi'me some more!
+              pull(in)
+            }
         }
       }
 
       override def onUpstreamFinish(): Unit = {
-        if (subSource.isDefined) {
-          subSource.get.complete()
-        } else {
-          materializeFlow()
+        subSource match {
+          case OptionVal.Some(s) => s.complete()
+          case OptionVal.None    => materializeFlow()
         }
       }
 
       override def onUpstreamFailure(ex: Throwable): Unit = {
-        if (subSource.isDefined) {
-          subSource.get.fail(ex)
-        } else {
-          //flow won't be materialized, so we have to complete the future with a failure indicating this
-          matPromise.failure(new NeverMaterializedException(ex))
-          super.onUpstreamFailure(ex)
+        subSource match {
+          case OptionVal.Some(s) => s.fail(ex)
+          case OptionVal.None    =>
+            //flow won't be materialized, so we have to complete the future with a failure indicating this
+            matPromise.failure(new NeverMaterializedException(ex))
+            super.onUpstreamFailure(ex)
         }
       }
 
       override def onPull(): Unit = {
-        if (subSink.isDefined) {
-          //delegate to subSink
-          subSink.get.pull()
-        } else if (accumulated.size < n) {
-          pull(in)
-        } else if (accumulated.size == n) { //corner case for n = 0, can be handled in FlowOps
-          materializeFlow()
+        subSink match {
+          case OptionVal.Some(s) =>
+            //delegate to subSink
+            s.pull()
+          case OptionVal.None if accumulated.size < n =>
+            pull(in)
+          case OptionVal.None if accumulated.size == n =>
+            //corner case for n = 0, can be handled in FlowOps
+            materializeFlow()
         }
       }
 
       override def onDownstreamFinish(cause: Throwable): Unit = {
-        if (subSink.isEmpty) {
-          downstreamCause = OptionVal.Some(cause)
-        } else {
-          subSink.get.cancel(cause)
+        subSink match {
+          case OptionVal.None    => downstreamCause = OptionVal.Some(cause)
+          case OptionVal.Some(s) => s.cancel(cause)
         }
       }
 
       def materializeFlow(): Unit =
         try {
           val prefix = accumulated.toVector
           accumulated.clear()
-          subSource = OptionVal.Some(new SubSourceOutlet[In]("subSource"))
-          subSource.get.setHandler {
+          subSource = OptionVal.Some(new SubSourceOutlet[In](s"${this}.subSource"))
+          subSource.x.setHandler {
```

Comment:
That's probably fast enough but maybe just let's create another `val` with the value and just use that one in here?

Like 

```scala
val theSubSource = new SubSourceOutlet[In](s"${this}.subSource")
this.subSource = OptionVal.Some(theSubSource)
theSubSource.setHandler
...
```